src/ddsrt/tests/dl.c

```diff
@@ -0,0 +1,224 @@
+
+#define TEST_ABORT_IF_NULL(var, msg) \
+do { \
+  if (var == NULL) { \
+    char buffer[256]; \
+    r = ddsrt_dlerror(buffer, sizeof(buffer)); \
+    CU_ASSERT_EQUAL_FATAL(r, DDS_RETCODE_OK); \
+    printf("\n%s", buffer); \
+    CU_FAIL_FATAL(msg); \
+  } \
+} while(0)
+
+
+/*
+ * Load a library.
+ */
+CU_Test(ddsrt_library, dlopen_path)
+{
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  printf("Absolute lib: %s\n", TEST_LIB_ABSOLUTE);
+  r = ddsrt_dlopen(TEST_LIB_ABSOLUTE, false, &l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  TEST_ABORT_IF_NULL(l, "ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlopen_file)
+{
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  r = ddsrt_dlopen(TEST_LIB_FILE, false, &l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  TEST_ABORT_IF_NULL(l, "ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlopen_name)
+{
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  r = ddsrt_dlopen(TEST_LIB_NAME, true, &l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  TEST_ABORT_IF_NULL(l, "ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlopen_unknown)
+{
+  char buffer[256];
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  r = ddsrt_dlopen("UnknownLib", false, &l);
+
+  CU_ASSERT_PTR_NULL_FATAL(l);
+
+  r = ddsrt_dlerror(buffer, sizeof(buffer));
+  CU_ASSERT_EQUAL_FATAL(r, DDS_RETCODE_OK);
+  printf("\n%s", buffer);
+}
+
+CU_Test(ddsrt_library, dlopen_null)
+{
+  char buffer[256];
+  dds_retcode_t r;
+  ddsrt_lib l;
+
+  r = ddsrt_dlopen(NULL, false, &l);
+  CU_ASSERT_EQUAL_FATAL(r, DDS_RETCODE_BAD_PARAMETER);
+  
+  printf("\n%s", buffer);
+}
+
+CU_Test(ddsrt_library, dlsym)
+{
+  dds_retcode_t r;
+  ddsrt_lib l;
+  void* f;
+
+  r = ddsrt_dlopen(TEST_LIB_NAME, true, &l);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  TEST_ABORT_IF_NULL(l, "ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlsym(l, "get_int", &f);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(f);
+  TEST_ABORT_IF_NULL(f, "ddsrt_dlsym(l, \"get_int\") failed.");
+
+  r = ddsrt_dlclose(l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+}
+
+CU_Test(ddsrt_library, dlsym_bad)
+{
+  dds_retcode_t r;
+  ddsrt_lib l;
+  void* f;
+
+  r = ddsrt_dlopen(TEST_LIB_NAME, true, &l);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_OK);
+  CU_ASSERT_PTR_NOT_NULL(l);
+  TEST_ABORT_IF_NULL(l,"ddsrt_dlopen() failed. Is the proper library path set?");
+
+  r = ddsrt_dlsym(l, NULL, &f);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_BAD_PARAMETER);
+
+  r = ddsrt_dlsym(NULL, "UnknownSym", &f);
+  CU_ASSERT_EQUAL(r, DDS_RETCODE_BAD_PARAMETER);
+
```

Comment:
I'd also expect a test that checks what happens when the symbol argument is NULL.